model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java

```diff
@@ -1777,55 +1763,34 @@ public void removeComponent(ComponentModel component) {
         if (c == null) return;
         session.users().preRemove(this, component);
         removeComponents(component.getId());
-        em.createNamedQuery("deleteComponentConfigByComponent").setParameter("component", c).executeUpdate();
-        em.remove(c);
+        getEntity().getComponents().remove(c);
     }
 
     @Override
     public void removeComponents(String parentId) {
-        TypedQuery<String> query = em.createNamedQuery("getComponentIdsByParent", String.class)
-                .setParameter("realm", realm)
-                .setParameter("parentId", parentId);
-        List<String> results = query.getResultList();
-        if (results.isEmpty()) return;
-        for (String id : results) {
-            session.users().preRemove(this, getComponent(id));
-        }
-        em.createNamedQuery("deleteComponentConfigByParent").setParameter("parentId", parentId).executeUpdate();
-        em.createNamedQuery("deleteComponentByParent").setParameter("parentId", parentId).executeUpdate();
-
+        getEntity().getComponents().stream()
+                .filter(c -> parentId.equals(c.getParentId()))
+                .forEach(c -> session.users().preRemove(this, getComponent(c.getId())));
```

Comment:
I thought about this version, but I suspect a ConcurrentModificationException if modifying the collection while it's being read by the stream. I've also thought about that pattern, but I feel that it may become cryptic we try to much to wrap everything around the stream. 

    @Override
    public void removeComponents(String parentId) {
        getEntity().getComponents().removeAll(
                getEntity().getComponents().stream()
                        .filter(c -> Objects.equals(parentId, c.getParentId()))
                        .map(c -> {
                            session.users().preRemove(this, getComponent(c.getId()));
                            return c;
                        })
                        .collect(Collectors.toList())
        );
    }